torch/autograd/profiler_legacy.py

```diff
@@ -186,7 +186,7 @@ def _get_record_key(record):
                     duplicate = (
                         prev_record.name() == record.name()
                         and prev_record.kind() == record.kind()
-                        and prev_record.node_id() == record.node_id()
+                        and prev_record.start_us() == record.start_us()
```

Comment:
See this comment. https://github.com/pytorch/pytorch/pull/62710#issuecomment-893203087

Reply:
This testing failure occurs in this branch right after Ilia's refactor. It serve as quick dirty fix toward legacy code. If you insist, I can do more investigation about it.

Reply:
YEs please, relying on the time stamp to identify duplicates can result in false positive matches.

Reply:
@ilia-cher do you know why your refactor would cause this?

Reply:
@ngimel 

https://github.com/pytorch/pytorch/blob/0507c316cd63cecd899849337d56fc38a8f62c2b/torch/autograd/profiler_legacy.py#L175

after this line, I added a print 
```py
print("node_id:", record.node_id(), "start_us:", record.start_us(), "name:", record.name())
```

trigger the test simply with 
```
pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_record_function_on_caller_rpc_async
```

<details>
  <summary>expand to see the stdout</summary>

```
node_id: 0 start_us: 1629782540852445.2 name: aten::add
node_id: 0 start_us: 1629782540852593.8 name: 
node_id: 0 start_us: 1629782540856163.8 name: __exit__
node_id: 0 start_us: 1629782540856210.5 name: 
node_id: 0 start_us: 1629782540691356.5 name: __start_profile
node_id: 0 start_us: 1629782540696474.5 name: defaults
node_id: 0 start_us: 1629782540696482.8 name: 
node_id: 0 start_us: 1629782540718015.0 name: defaults
node_id: 0 start_us: 1629782540718018.0 name: 
node_id: 0 start_us: 1629782540720749.8 name: defaults
node_id: 0 start_us: 1629782540720753.0 name: 
node_id: 0 start_us: 1629782540746427.5 name: defaults
node_id: 0 start_us: 1629782540746430.5 name: 
node_id: 0 start_us: 1629782540748283.5 name: defaults
node_id: 0 start_us: 1629782540748286.5 name: 
node_id: 0 start_us: 1629782540754889.5 name: defaults
node_id: 0 start_us: 1629782540754892.5 name: 
node_id: 0 start_us: 1629782540838905.8 name: record_function_on_caller_rpc_async
node_id: 0 start_us: 1629782540838916.5 name: aten::ones
node_id: 0 start_us: 1629782540838946.8 name: aten::empty
node_id: 0 start_us: 1629782540838961.8 name: 
node_id: 0 start_us: 1629782540838975.2 name: aten::fill_
node_id: 0 start_us: 1629782540839018.5 name: 
node_id: 0 start_us: 1629782540839022.2 name: 
node_id: 0 start_us: 1629782540839775.0 name: __enter__
node_id: 0 start_us: 1629782540839782.8 name: foo
node_id: 0 start_us: 1629782540839795.0 name: 
node_id: 0 start_us: 1629782540839840.5 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)
node_id: 0 start_us: 1629782540840022.2 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)
node_id: 0 start_us: 1629782540850260.0 name: 
node_id: 0 start_us: 1629782540851111.2 name: 
node_id: 0 start_us: 1629782540856195.8 name: 
node_id: 0 start_us: 1629782540856257.5 name: 
node_id: 0 start_us: 1629782540856288.5 name: __stop_profile
node_id: 1 start_us: 1629782540840774.8 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: __start_profile
node_id: 1 start_us: 1629782540849054.5 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: script_add_ones
node_id: 1 start_us: 1629782540849112.0 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: aten::add
node_id: 1 start_us: 1629782540849179.0 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: 
node_id: 1 start_us: 1629782540849251.8 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: 
node_id: 1 start_us: 1629782540849292.5 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: __stop_profile
node_id: 1 start_us: 1629782540840564.5 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: __start_profile
node_id: 1 start_us: 1629782540842807.2 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: defaults
node_id: 1 start_us: 1629782540842812.2 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: 
node_id: 1 start_us: 1629782540846464.8 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: aten::ones
node_id: 1 start_us: 1629782540846489.2 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: aten::empty
node_id: 1 start_us: 1629782540846500.5 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: 
node_id: 1 start_us: 1629782540846510.5 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: aten::fill_
node_id: 1 start_us: 1629782540846545.0 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: 
node_id: 1 start_us: 1629782540846548.8 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: 
node_id: 1 start_us: 1629782540849036.2 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: script_add_ones
node_id: 1 start_us: 1629782540849077.5 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: aten::add
node_id: 1 start_us: 1629782540849176.2 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: 
node_id: 1 start_us: 1629782540849231.2 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: 
node_id: 1 start_us: 1629782540849271.0 name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)#remote_op: __stop_profile
```
</details>

as you can see, most record has duplicated `node_id`...

Reply:
OK, it seems that I have some misunderstanding of `node_id` in this case. It is not the ID of the profiling event. It is the machine node id that emit the event. I will revert the change of this line.

Reply:
@ngimel OK, I nailed it. I checkout 773a8eede437de972311b21824694e4fb120f6a2, the commit of @ilia-cher 's refactor. With the previous print hack to inspect into the deduplication process

```
...
node_id: 0 kind: push name: aten::empty
node_id: 0 kind: push name: aten::zeros
node_id: 0 kind: push name: aten::empty
node_id: 0 kind: push name: aten::zero_
node_id: 0 kind: push name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)
node_id: 0 kind: push name: aten::empty
node_id: 0 kind: push name: aten::zeros
node_id: 0 kind: push name: aten::empty
node_id: 0 kind: push name: aten::zero_
node_id: 0 kind: push name: rpc_async_jit#__torch__.torch.testing._internal.distributed.rpc.jit.rpc_test.script_add_ones(worker0 -> worker1)
node_id: 0 kind: push name: aten::empty
...
```

The reason the `...script_add_ones...`s were not deduplicated is because original `cpp_custom_type_hack` introduce `aten::empty`, `aten::zeros`, `aten::empty`, `aten::zero_` events between two `...script_add_ones...`s. Now, without the hack, two `...script_add_ones...`s appears **consecutive**, their `name`, `kind`, `node_id` are the same, and then deduplicated.

Reply:
@ngimel I update the workaround. Instead of trying to fix the deduplication logic in `profiler_legacy.py`. Instead, I added a dummy operation between the two `rpc_async` in the test case. So that they are not consecutive anymore. How to avoid false-positively classify  consecutive `...script_add_ones...`s as duplicated is left to pytorch team.